File: slang-plugin/src/test/java/org/sonarsource/slang/plugin/StatementsVisitorTest.java
/*
 * SonarSource SLang
 * Copyright (C) 2018-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarsource.slang.plugin;

import org.junit.jupiter.api.Test;
import org.sonarsource.slang.api.Tree;
import org.sonarsource.slang.parser.SLangConverter;

import static org.assertj.core.api.Assertions.assertThat;

class StatementsVisitorTest {

  private static int statements(String content) {
    Tree root = new SLangConverter().parse(content);
    return new StatementsVisitor().statements(root);
  }

  @Test
  void should_count_top_level_without_natives_and_blocks() throws Exception {
    String content = "" +
      "package abc;" +
      "import xyz;" +
      "native[] { };" +
      "foo;" + // +1
      "class A{};" +
      "if (a) {};" + // +1
      "fun foo() {};" +
      "{};";
    assertThat(statements(content)).isEqualTo(2);
  }

  @Test
  void should_count_statements_inside_blocks() throws Exception {
    String content = "fun foo() {" +
      "native[] { };" + // +1
      "foo;" + // +1
      "if (a) { foo; bar; };" + // +1 +1 +1
      "class A{};" +
      "fun bar(){};" +
      "};" +
      "{ 2; 3; };"; // +1 +1
    assertThat(statements(content)).isEqualTo(7);
  }

}

File: sonar-kotlin-plugin/src/test/java/org/sonarsource/kotlin/plugin/StatementsVisitorTest.kt
/*
 * SonarSource SLang
 * Copyright (C) 2018-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonarsource.kotlin.plugin

import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import org.sonarsource.kotlin.converter.KotlinConverter

internal class StatementsVisitorTest {

    @Test
    fun `should count top level without declarations and blocks`() {
        val content = """
            package abc
            import xyz
            
            class A{}
            fun bar() = if (a) println()
            fun foo() {}
            """.trimIndent()
        Assertions.assertThat(statements(content)).isEqualTo(2)
    }

    @Test
    fun `should count statements inside blocks`() {
        val content = """
            fun foo(a: Boolean) {
              foo()
              if (a) { 
                foo()
                bar()
              }
              class A{}
              fun bar(){}
            }
            fun f() {
              var a = 2
              var b = 3 
            }""".trimIndent()
        Assertions.assertThat(statements(content)).isEqualTo(6)
    }
}

private fun statements(content: String): Int {
    val root = KotlinConverter(emptyList()).parse(content)
    val statementsVisitor = StatementsVisitor()
    root.psiFile.accept(statementsVisitor)
    return statementsVisitor.statements
}
